### ISSUE TITLE
오버피팅일 경우 어떻게 대처해야 할까요?

### ISSUE THREAD

**Issue by jarammm** [AInsteins/AI-Tech-Interview#3]
[1회차] 2022년 8월 23일 화요일

**Comment by jarammm** [comment]
오버피팅은 실험하고 있는 특정 데이터를 필요 이상으로 학습한 상태로, 일반화 성능이 낮아지는 문제를 초래합니다.

1. **데이터의 수 증가(augmentation을 통해 가능)**
2. 데이터에 노이즈 추가 : outlier에 robust하도록 합니다.
3. feature를 줄이거나, layer를 줄여 모델의 복잡도를 낮춥니다.
4. 배치 정규화(batch normalization) : mini batch 별로 분산과 표준편차를 구해 분포를 조정합니다. 이를 통해 가중치의 초깃값에 의존하지 않고 **강제로** 활성화 값을 적절히 분포되도록 합니다.
5. 혹은 dropout을 통해 각 레이어의 뉴런 간 연결을 랜덤하게 끊음으로써 과적합을 막기도 합니다.
6. 가중치 감소 : 손실 함수에 L1 loss, 혹은 L2 loss 값을 더하여, 가중치에 비례하는 패널티를 부과하여 오버피팅을 억제합니다.
    1. L1 규제 : 불필요한 가중치는 0으로 만들어버림 → robust하진 않음
    2. L2 규제 : 불필요한 가중치는 0에 가깝게 만들어버림

**Comment by import-max** [comment]
- 오버피팅이란 일반적으로 학습 데이터에만 과도하게 적응이 되어 실험 데이터에서는 안 좋은 성능이 나타나는 경우인데, 이를 해결하기 위해서는 훈련 데이터와 학습 데이터의 수를 맞춰줄 수 있다. 학습 데이터가 적고 실험 데이터가 많은 경우라면, 학습 데이터의 양을 늘려줄 수 있다. 
혹은 정규화를 통해 가중치의 매개변수 값을 조정할 수 있는데, 이 값이 클 수록 오버피팅이 발생하는 확률도 크기 때문에 가중치가 크다면 패널티를 크게 줌으로써 정규화를 해줄 수 있다.
early stop을 통해 training epoch를 돌 때마다의 loss를 체크하여 training loss 와 validation이 반비례하는 시점에서 학습을 종료할 수도 있다.
- 또 일부러 데이터에 노이즈를 준다든가, Ground Truth의 smoothing하는 방법도 있다.
- Dropout을 통해 layer를 쌓을 때마다 학습되는 신경망을 임의로 drop시키고 일부만 학습시킴으로써 전체적으로 골고루 training을 하면 overfitting을 막으면서도 ensemble하는 효과를 낼 수 있다.

**Comment by import-max** [comment]
[https://saige.ai/ko/a-peek-into-automatic-data-augmentation-by-policy-searching-2/](url)

**Comment by kimyuuuuuu** [comment]
- 오버피팅이란 학습데이터를과하게 학습하는 것으로 학습데이터의 오차는 줄고 실제데이터의 오차는 늘어나는 경우이다.
- 이 경우 데이터에 인위적으로 변화를 주어 훈련데이터를 증가시키는 Data argumentation
- train set의 정확성은 올라가나(= loss는 내려가나 ) validation set은 멈추거나 내려가는 시점에서 epoch 수가 완료되지 않아도 멈추는 Early stopping
- train 과정에서 뉴런 몇개를 쉬게하는 Drop out : 앙상블 효과
- 모델의 복잡도를 낮추는 Regularizaton :
  -  L1 : w값 자체를 줄이는게 아니라 상수값을 뺌 -> 0에 수렴. 
  -  L2 : 곱해진 값에 의해 w에 작아짐

**Comment by miinkang** [comment]
오버피팅은 데이터의 양이 부족하거나 파라미터의 개수가 많을 때 발생할 수 있습니다. 

따라서 데이터의 양을 늘리는 방법을 취할 수 있습니다. 
모델의 레이어, 노드의 수를 줄여서 모델의 복잡도를 낮추어서 파라미터의 개수를 줄일 수 있습니다. 
dropout을 이용해 네트워크의 일부만 학습시키거나 batch normalization을 사용할 수 있습니다.

**Comment by jarammm** [comment]
@import-max 질문) 데이터의 수가 작은 경우 augmentation 진행 시, 노이즈의 영향을 강하게 받아 오히려 학습 성능이 나빠질 수 있는데 이 때 어떻게 하나?
